Define BFloat16 intrinsics

eopXD opened the pull request:
NOTE: This PR currently is based upon #292 

This pull request defines intrinsics for vector bfloat16 type manipulations and vector instructions that are introduced by the RISC-V BFloat16 instruction set extensions (`zvfbfmin` and `zvfbfwma`) [0].

Intrinsics introduced for vector bfloat16 type manipulation:
- (Non-segment/Segment) loads and stores around the vector bfloat16 types.
- Utility intrinsics like `vcreate`, `vundefined`, ... etc (you may find the exact list under generated prototypes and its declarations under `bfloat16_inst.py`)

New intrinsics are added to expose new instructions introduced:
- `vfwmaccbf16.vv vd, vs1, vs2, vm`
- `vfwmaccbf16.vf vd, rs1, vs2, vm`
- `vfncvtbf16.f.f.w vd, vs2, vm`
- `vfwcvtbf16.f.f.v vd, vs2, vm`

[0] https://github.com/riscv/riscv-bfloat16

rofirrim reviewed (commented):
We are using `__bf16` for the scalar type, can be add a reference to some specification about this name?

Other than that, everything else LGTM.

dzaima commented:
Shouldn't `__riscv_vreinterpret_v_ui16m1_bf16m1` be `__riscv_vreinterpret_v_u16m1_bf16m1` (i.e. `u16` instead of `ui16`)?

kito-cheng commented:
https://github.com/riscv-non-isa/riscv-elf-psabi-doc/blob/master/riscv-cc.adoc#cc-type-sizes-and-alignments

`__bf16` has used in psABI, but we may also need to add that to c-api-doc https://github.com/riscv-non-isa/riscv-c-api-doc/blob/master/riscv-c-api.md for more complete documentation for that type.

eopXD commented:
Rebase and addressed comments from @dzaima, thank you for pointing this out.
@kito-cheng I'll leave it to you to add some descriptions somewhere in the C api document and we'll then cite them here.

rofirrim reviewed (approved):
LGTM. Thanks @eopXD !

tclin914 commented on auto-generated/bfloat16/llvm-api-tests/vfncvtbf16.c:
```diff
@@ -0,0 +1,97 @@
+// REQUIRES: riscv-registered-target
+// RUN: %clang_cc1 -triple riscv64 -target-feature +v -target-feature +zfh \
+// RUN:   -target-feature +experimental-zvfh -disable-O0-optnone  \
```
Does it should be changed to `+experimental-zvfbfmin`

tclin914 commented on auto-generated/bfloat16/api-testing/vlmul_ext_v.c:
```diff
@@ -0,0 +1,62 @@
+#include <riscv_vector.h>
+#include <stdint.h>
+
+vbfloat16mf2_t test_vlmul_ext_v_b16mf4_b16mf2(vbfloat16mf4_t value) {
```
In vlmul_ext_v.c and vlmul_trunc_v.c, b16->bf16.
rvv_intrinsic_gen/templates/misc_op_template.py:109 `args["TYPE1"] = args["TYPE"][0]`

4vtomat commented:
I'll take over and proceed this patch!

topperc commented on doc/references.adoc:
```diff
@@ -0,0 +1,67 @@
+== References
```
Oh this was separated out of rvv-intrinsic-spec.adoc. I just hadn't gotten there yet.

topperc reviewed (approved):
LGTM
